src/source/test_waveform.rs

```diff
@@ -0,0 +1,196 @@
+//! Generator sources for various periodic test waveforms.
+//!
+//! This module provides several periodic, deterministic waveforms for testing other sources and
+//! for simple additive sound synthesis. Every source is monoaural and in the codomain `[-1.0f32,
+//! 1.0f32]` 
+//!
+//! # Example
+//!
+//! ```
+//! use rodio::source::{TestWaveform,TestWaveformFunction};
+//!
+//! let tone = TestWaveform::new(cpal::SampleRate(48000), 440.0, TestWaveformFunction::Sine);
+//! ```
+use std::f32::consts::TAU;
+use std::time::Duration;
+
+use super::SeekError;
+use crate::Source;
+
+/// Test waveform functions.
+#[derive(Clone, Debug)]
+pub enum TestWaveformFunction {
+    /// A sinusoidal waveform.
+    Sine,
+    /// A triangle wave.
+    Triangle,
+    /// A square wave, rising edge at t=0.
+    Square,
+    /// A rising swatooth wave.
+    Sawtooth,
+}
+
+impl TestWaveformFunction {
+    /// Create a single sample for the given waveform
+    #[inline]
+    fn render(&self, i: u64, period: f32) -> f32 {
+        let i_div_p: f32 = i as f32 / period;
+
+        match self {
+            Self::Sine => (TAU * i_div_p).sin(),
+            Self::Triangle => 04.0f32 * (i_div_p - (i_div_p + 0.5f32).floor()).abs() - 1f32,
+            Self::Square => {
+                if i_div_p % 1.0f32 < 0.5f32 {
+                    1.0f32
+                } else {
+                    -1.0f32
+                }
+            }
+            Self::Sawtooth => 2.0f32 * (i_div_p - (i_div_p + 0.5f32).floor()),
+        }
+    }
+}
+
+/// An infinite source that produces one of a selection of test waveforms.
+#[derive(Clone, Debug)]
+pub struct TestWaveform {
+    sample_rate: cpal::SampleRate,
+    period: f32,
+    f: TestWaveformFunction,
+    i: u64,
+}
+
+impl TestWaveform {
+    /// Create a new `TestWaveform` object that generates an endless waveform
+    /// `f`.
+    #[inline]
+    pub fn new(
+        sample_rate: cpal::SampleRate,
+        frequency: f32,
+        f: TestWaveformFunction,
+    ) -> TestWaveform {
+        let period = sample_rate.0 as f32 / frequency;
+        TestWaveform {
+            sample_rate,
+            period,
+            f,
+            i: 0,
+        }
+    }
+}
+
+impl Iterator for TestWaveform {
+    type Item = f32;
+
+    #[inline]
+    fn next(&mut self) -> Option<f32> {
+        let this_i = self.i;
+        self.i += 1;
+        Some(self.f.render(this_i, self.period))
+    }
+}
+
+impl Source for TestWaveform {
+    #[inline]
+    fn current_frame_len(&self) -> Option<usize> {
+        None
+    }
+
+    #[inline]
+    fn channels(&self) -> u16 {
+        1
+    }
+
+    #[inline]
+    fn sample_rate(&self) -> u32 {
+        self.sample_rate.0
+    }
+
+    #[inline]
+    fn total_duration(&self) -> Option<Duration> {
+        None
+    }
+
+    #[inline]
+    fn try_seek(&mut self, duration: Duration) -> Result<(), SeekError> {
+        self.i = (self.sample_rate.0 as f32 * duration.as_secs_f32()) as u64;
```

Comment:
Does it make sense to implement `try_seek`? Seek was meant as a rough way to go to a  point in time. The duration is measured from the start of playback. 

You seek in a music file a podcast etc usually based on imprecise instructions from the user (go forward 5s go to 13m:23s). If you need precise narrow adjustments to playback (35ms forward) skipping samples seems the better approach (see [the skip duration method which is part of Source](https://docs.rs/rodio/0.19.0/rodio/source/trait.Source.html#method.skip_duration)).

Reply:
Is there a *precise* seeking method? I noticed this when I was chaining together sources for crossfades, sometimes you need to split a source/skip a source to a precise sample.

Reply:
unfortunately not, the seek implementation is left over to the decoder. The decoder can be really precise, but it could also be off for a bit. Think off by milliseconds not seconds.

Reply:
Jus getting back into this, would you like me to delete this from the implementation?

Reply:
nah its okay, its not that complex. Fine to keep it in.

Reply:
I was on the fence but I thought about it and I really think we do not want ppl to rely on try_seek for such precise 'seeking'. Skip samples is the correct approach for that. Do you agree?